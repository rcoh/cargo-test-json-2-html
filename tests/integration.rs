use cargo_test_json_2_html::{Config, SourceLinker, convert_to_html};
use std::fs;
use proptest::prelude::*;

#[derive(Debug)]
struct TestLinker;

impl SourceLinker for TestLinker {
    fn link(&self, file: &str, line: u32) -> Option<String> {
        Some(format!(
            "https://github.com/test/repo/blob/main/{}#L{}",
            file, line
        ))
    }
}

proptest! {
    #[test]
    fn convert_to_html_never_panics(input in ".*") {
        let config = Config::default();
        let _ = convert_to_html(&input, config);
    }
}

#[test]
fn test_xss_protection() {
    let input = fs::read_to_string("test-data/xss-protection.json").unwrap();
    let config = Config::default();
    let html = convert_to_html(&input, config);

    // Should not contain unescaped script tags
    assert!(!html.contains("<script>alert('xss')</script>"));
    // Should contain escaped version
    assert!(html.contains("&lt;script&gt;"));

    insta::assert_snapshot!(html);
}

#[test]
fn test_source_linking() {
    let input = fs::read_to_string("test-data/source-linking.json").unwrap();
    let config = Config::builder().source_linker(TestLinker).build();
    let html = convert_to_html(&input, config);

    insta::assert_snapshot!(html);
}

#[test]
fn test_real_cargo_output() {
    let input = fs::read_to_string("test-data/basic.json").unwrap();
    let config = Config::default();
    let html = convert_to_html(&input, config);

    insta::assert_snapshot!(html);
}

#[test]
fn test_mixed_output_with_compilation() {
    let input = fs::read_to_string("test-data/mixed-compilation.json").unwrap();
    let config = Config::default();
    let html = convert_to_html(&input, config);

    insta::assert_snapshot!(html);
}

#[test]
fn test_error_handling() {
    let input = fs::read_to_string("test-data/error-handling.json").unwrap();
    let config = Config::default();
    let html = convert_to_html(&input, config);

    insta::assert_snapshot!(html);
}
