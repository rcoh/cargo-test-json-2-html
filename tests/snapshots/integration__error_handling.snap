---
source: tests/integration.rs
expression: html
---
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Cargo Test Report</title>
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 20px; background: #f5f5f5; }
        .container { max-width: 1200px; margin: 0 auto; background: white; border-radius: 8px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
        .header { background: #2563eb; color: white; padding: 20px; border-radius: 8px 8px 0 0; }
        .content { padding: 20px; }
        
        /* Improved summary design */
        .summary { background: #f8fafc; border: 1px solid #e2e8f0; border-radius: 8px; padding: 20px; margin-bottom: 30px; }
        .summary-title { font-size: 16px; font-weight: 600; color: #374151; margin-bottom: 15px; }
        .stats-row { display: flex; align-items: center; gap: 20px; flex-wrap: wrap; }
        .stat-item { display: flex; align-items: center; gap: 8px; }
        .stat-item a { text-decoration: none; color: inherit; display: flex; align-items: center; gap: 8px; }
        .stat-item a:hover { opacity: 0.8; }
        .stat-icon { width: 16px; height: 16px; border-radius: 50%; }
        .stat-icon.passed { background: #10b981; }
        .stat-icon.failed { background: #ef4444; }
        .stat-icon.ignored { background: #f59e0b; }
        .stat-text { font-size: 14px; color: #374151; }
        .stat-number { font-weight: 600; }
        .duration { margin-left: auto; font-size: 14px; color: #6b7280; }
        
        .section { margin-bottom: 30px; }
        .section-title { font-size: 18px; font-weight: 600; margin-bottom: 15px; color: #1e293b; }
        .test-item { background: #f8fafc; border: 1px solid #e2e8f0; border-radius: 6px; margin-bottom: 10px; }
        
        /* Improved clickable indication */
        .test-header { padding: 15px; display: flex; justify-content: space-between; align-items: center; }
        .test-header.clickable { cursor: pointer; transition: background-color 0.2s; }
        .test-header.clickable:hover { background: #f1f5f9; }
        
        .test-right { display: flex; align-items: center; gap: 10px; }
        .view-output { font-size: 12px; color: #6b7280; transition: opacity 0.2s; }
        .view-output.no-output { opacity: 0.6; }
        .test-header:hover .view-output { color: #374151; }
        .test-header:hover .view-output.no-output { opacity: 1; }
        
        .test-name { font-family: 'Monaco', 'Menlo', monospace; font-size: 14px; }
        .test-status { padding: 4px 8px; border-radius: 4px; font-size: 12px; font-weight: 500; }
        .test-status.passed { background: #dcfce7; color: #166534; }
        .test-status.failed { background: #fecaca; color: #991b1b; }
        .test-status.ignored { background: #fef3c7; color: #92400e; }
        .test-details { padding: 0 15px 15px; border-top: 1px solid #e2e8f0; display: none; }
        .test-details.show { display: block; }
        .test-output { background: #1e293b; color: #e2e8f0; padding: 15px; border-radius: 4px; font-family: 'Monaco', 'Menlo', monospace; font-size: 13px; white-space: pre-wrap; overflow-x: auto; }
        .test-output a { color: #60a5fa; text-decoration: underline; }
        .test-output a:hover { color: #93c5fd; }
        .errors { background: #fef2f2; border: 1px solid #fecaca; border-radius: 6px; padding: 15px; }
        .error-item { margin-bottom: 10px; font-family: 'Monaco', 'Menlo', monospace; font-size: 13px; }
        .raw-lines { background: #f8fafc; border: 1px solid #e2e8f0; border-radius: 6px; padding: 15px; }
        .raw-line { font-family: 'Monaco', 'Menlo', monospace; font-size: 13px; margin-bottom: 5px; }
        
        @media (max-width: 600px) {
            .test-status { display: none; }
        }
    </style>
    <script>
        function toggleDetails(element) {
            const details = element.nextElementSibling;
            details.classList.toggle('show');
        }
    </script>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Cargo Test Report</h1>
        </div>
        <div class="content">
            <div class="summary">
                <div class="summary-title">Test Results Summary</div>
                <div class="stats-row">
                    <div class="stat-item">
                        <a href="#passed-tests">
                            <div class="stat-icon passed"></div>
                            <div class="stat-text"><span class="stat-number">1</span> passed</div>
                        </a>
                    </div>
                    <div class="stat-item">
                        <a href="#failed-tests">
                            <div class="stat-icon failed"></div>
                            <div class="stat-text"><span class="stat-number">0</span> failed</div>
                        </a>
                    </div>
                    <div class="stat-item">
                        <a href="#ignored-tests">
                            <div class="stat-icon ignored"></div>
                            <div class="stat-text"><span class="stat-number">0</span> ignored</div>
                        </a>
                    </div>
                </div>
            </div>


            <div class="section">
                <div class="section-title" id="passed-tests">‚úÖ Passed Tests</div>
                <div class="test-item">
                    <div class="test-header">
                        <span class="test-name">valid_test</span>
                        <div class="test-right">
                            <span class="view-output no-output">No output</span>
                            <span class="test-status passed">PASSED</span>
                        </div>
                    </div>
                </div>
            </div>


            <div class="section">
                <div class="section-title">‚ö†Ô∏è Parsing Errors</div>
                <div class="errors">
                    <div class="error-item">Failed to parse JSON: expected value at line 1 column 16 - Line: { &quot;malformed&quot;: json }</div>
                </div>
            </div>

            <div class="section">
                <div class="section-title">üìÑ Raw Output</div>
                <div class="raw-lines">
                    <div class="raw-line">invalid json line</div>
                </div>
            </div>
        </div>
    </div>
</body>
</html>
